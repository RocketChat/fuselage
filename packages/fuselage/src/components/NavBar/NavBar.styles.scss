@use '../../styles/colors.scss';
@use '../../styles/lengths.scss';

.rcx-navbar {
  display: flex;
  justify-content: space-between;
  align-items: center;

  width: 100%;
  padding: lengths.padding(8) lengths.padding(16);

  border-bottom: lengths.border-width(default) solid colors.stroke(light);

  background-color: colors.surface(sidebar);

  &-section {
    display: flex;
    align-items: center;
  }

  &-divider {
    border-color: colors.stroke(medium);
  }
}
