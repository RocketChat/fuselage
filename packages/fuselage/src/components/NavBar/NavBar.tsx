import type { HTMLAttributes } from 'react';

export const NavBar = (props: HTMLAttributes<HTMLElement>) => (
  <nav className='rcx-navbar' {...props} />
);
