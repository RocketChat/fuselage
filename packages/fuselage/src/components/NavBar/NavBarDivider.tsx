import type { ComponentProps } from 'react';

import { Divider } from '../Divider';

export const NavBarDivider = (props: ComponentProps<typeof Divider>) => (
  <Divider rcx-navbar-divider vertical {...props} />
);
