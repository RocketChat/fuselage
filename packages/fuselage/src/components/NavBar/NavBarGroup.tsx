import type { ComponentProps } from 'react';
import { forwardRef } from 'react';

import { ButtonGroup } from '../ButtonGroup';

export const NavBarGroup = forwardRef<
  HTMLDivElement,
  ComponentProps<typeof ButtonGroup>
>((props, ref) => (
  <ButtonGroup className='rcx-navbar-group' ref={ref} {...props} />
));
