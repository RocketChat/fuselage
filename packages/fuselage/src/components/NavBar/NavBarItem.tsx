import type { ComponentProps, HTMLAttributes } from 'react';
import { forwardRef } from 'react';

import { appendClassName } from '../../helpers/appendClassName';
import { patchChildren } from '../../helpers/patchChildren';
import { IconButton } from '../Button';

type NavbarItemProps = Partial<ComponentProps<typeof IconButton>>;

export const NavBarItem = forwardRef<
  HTMLElement,
  HTMLAttributes<HTMLElement> & NavbarItemProps
>(function NavBarItem({ children, icon, ...props }, ref) {
  return (
    <>
      {icon ? (
        <IconButton ref={ref} rcx-navbar-item icon={icon} small {...props} />
      ) : (
        patchChildren(
          children,
          (childProps: { className: string | string[] }) => ({
            className: appendClassName(childProps.className, 'rcx-navbar-item'),
            ref,
            ...props,
          }),
        )
      )}
    </>
  );
});
