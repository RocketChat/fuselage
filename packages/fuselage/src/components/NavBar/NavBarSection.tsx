import type { HTMLAttributes } from 'react';
import React from 'react';

export const NavBarSection = (props: HTMLAttributes<HTMLSpanElement>) => (
  <span className='rcx-navbar-section' {...props} />
);
