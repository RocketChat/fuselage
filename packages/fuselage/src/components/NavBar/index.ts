export * from './NavBar';
export * from './NavBarGroup';
export * from './NavBarItem';
export * from './NavBarDivider';
export * from './NavBarSection';
