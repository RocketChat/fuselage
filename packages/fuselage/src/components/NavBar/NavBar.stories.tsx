import type { StoryFn, Meta } from '@storybook/react';
import type { ComponentType } from 'react';

import { Avatar } from '../Avatar';
import Box from '../Box';
import { MenuV2 as Menu, MenuItem, MenuSection } from '../Menu';
import { MenuDisplayExample } from '../Menu/V2/Menu.stories';
import { avatarUrl } from '../Message/helpers';
import { NavBar } from './NavBar';
import { NavBarDivider } from './NavBarDivider';
import { NavBarGroup } from './NavBarGroup';
import { NavBarItem } from './NavBarItem';
import { NavBarSection } from './NavBarSection';

export default {
  title: 'Navigation/Navbar',
  component: NavBar,
  subcomponents: {
    NavBarItem: NavBarItem as ComponentType<any>,
    NavBarSection: NavBarSection as ComponentType<any>,
    NavBarDivider: NavBarDivider as ComponentType<any>,
    NavBarGroup: NavBarGroup as ComponentType<any>,
  },
  decorators: [
    (Story) => (
      <Box display='flex'>
        <Story />
      </Box>
    ),
  ],
} satisfies Meta<typeof NavBar>;

const Template: StoryFn<typeof NavBar> = (args) => (
  <NavBar {...args}>
    <NavBarSection>
      <NavBarGroup role='toolbar'>
        <NavBarItem icon='home' title='home' />
        <NavBarItem icon='store' title='marketplace' />
        <NavBarItem icon='notebook-hashtag' title='directory' />
        <NavBarItem icon='document-eye' title='audit' />
        <NavBarItem icon='address-book' title='contacts' />
      </NavBarGroup>
      <NavBarDivider />
      <NavBarGroup role='toolbar'>
        <NavBarItem>
          <MenuDisplayExample small icon='sort' title='sort' />
        </NavBarItem>
        <NavBarItem icon='magnifier' title='search' />
        <Menu icon='pencil-box' small title='create new'>
          <MenuSection title='Create new'>
            <MenuItem key='direct-message'>Direct message</MenuItem>
            <MenuItem key='discussion'>Discussion</MenuItem>
            <MenuItem key='channel'>Channel</MenuItem>
            <MenuItem key='team'>Team</MenuItem>
          </MenuSection>
        </Menu>
      </NavBarGroup>
      <NavBarDivider />
      <NavBarGroup role='toolbar'>
        <NavBarItem icon='dialpad' title='dialpad' />
        <NavBarItem icon='live' title='live' />
        <NavBarItem icon='phone-disabled' title='phone-disabled' />
        <NavBarItem icon='message-disabled' title='message-disabled' />
      </NavBarGroup>
    </NavBarSection>
    <NavBarSection>
      <NavBarGroup>
        <NavBarItem icon='cog' title='preferences' />
        <NavBarItem>
          <Menu
            small
            icon={<Avatar url={avatarUrl} alt='user avatar' size='x28' />}
            title='profile'
          >
            <MenuItem key='1'>Profile</MenuItem>
            <MenuItem key='2'>Chats</MenuItem>
            <MenuItem key='3'>Settings</MenuItem>
          </Menu>
        </NavBarItem>
      </NavBarGroup>
    </NavBarSection>
  </NavBar>
);

export const Default: StoryFn<typeof NavBar> = Template.bind({});
