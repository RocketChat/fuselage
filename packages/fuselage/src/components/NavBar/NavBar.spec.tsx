import { composeStories } from '@storybook/testing-react';
import { render } from '@testing-library/react';
import { axe } from 'jest-axe';
import React from 'react';
import { SSRProvider } from 'react-aria';

import * as stories from './NavBar.stories';

const testCases = Object.values(composeStories(stories)).map((Story) => [
  Story.storyName || 'Story',
  Story,
]);

describe('[NavBar Component]', () => {
  test.each(testCases)(
    `renders %s without crashing`,
    async (_storyname, Story) => {
      const tree = render(<Story />, {
        wrapper: ({ children }) => <SSRProvider>{children}</SSRProvider>,
      });
      expect(tree.baseElement).toMatchSnapshot();
    }
  );

  test.each(testCases)(
    '%s should have no a11y violations',
    async (_storyname, Story) => {
      const { container } = render(<Story />);

      const results = await axe(container);
      expect(results).toHaveNoViolations();
    }
  );
});
