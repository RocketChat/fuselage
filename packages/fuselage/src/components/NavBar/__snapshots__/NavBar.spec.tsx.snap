// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[NavBar Component] renders Default without crashing 1`] = `
<body>
  <div>
    <div
      class="rcx-box rcx-box--full rcx-css-1qvl0ud"
    >
      <nav
        class="rcx-navbar"
      >
        <span
          class="rcx-navbar-section"
        >
          <div
            class="rcx-button-group rcx-button-group--align-start rcx-navbar-group"
            role="toolbar"
          >
            <button
              class="rcx-box rcx-box--full rcx-navbar-item rcx-button--small-square rcx-button--square rcx-button--icon rcx-button  rcx-button-group__item"
              title="home"
              type="button"
            >
              <i
                aria-hidden="true"
                class="rcx-box rcx-box--full rcx-icon--name-home rcx-icon rcx-css-4pvxx3"
              >
                
              </i>
            </button>
            <button
              class="rcx-box rcx-box--full rcx-navbar-item rcx-button--small-square rcx-button--square rcx-button--icon rcx-button  rcx-button-group__item"
              title="marketplace"
              type="button"
            >
              <i
                aria-hidden="true"
                class="rcx-box rcx-box--full rcx-icon--name-store rcx-icon rcx-css-4pvxx3"
              >
                
              </i>
            </button>
            <button
              class="rcx-box rcx-box--full rcx-navbar-item rcx-button--small-square rcx-button--square rcx-button--icon rcx-button  rcx-button-group__item"
              title="directory"
              type="button"
            >
              <i
                aria-hidden="true"
                class="rcx-box rcx-box--full rcx-icon--name-notebook-hashtag rcx-icon rcx-css-4pvxx3"
              >
                
              </i>
            </button>
            <button
              class="rcx-box rcx-box--full rcx-navbar-item rcx-button--small-square rcx-button--square rcx-button--icon rcx-button  rcx-button-group__item"
              title="audit"
              type="button"
            >
              <i
                aria-hidden="true"
                class="rcx-box rcx-box--full rcx-icon--name-document-eye rcx-icon rcx-css-4pvxx3"
              >
                
              </i>
            </button>
            <button
              class="rcx-box rcx-box--full rcx-navbar-item rcx-button--small-square rcx-button--square rcx-button--icon rcx-button  rcx-button-group__item"
              title="contacts"
              type="button"
            >
              <i
                aria-hidden="true"
                class="rcx-box rcx-box--full rcx-icon--name-address-book rcx-icon rcx-css-4pvxx3"
              >
                
              </i>
            </button>
          </div>
          <hr
            class="rcx-box rcx-box--full rcx-navbar-divider rcx-divider--vertical rcx-divider"
          />
          <div
            class="rcx-button-group rcx-button-group--align-start rcx-navbar-group"
            role="toolbar"
          >
            <button
              aria-expanded="false"
              aria-haspopup="true"
              class="rcx-box rcx-box--full rcx-button--small-square rcx-button--square rcx-button--icon rcx-button  rcx-button-group__item"
              id="react-aria-1"
              title="sort"
              type="button"
            >
              <i
                aria-hidden="true"
                class="rcx-box rcx-box--full rcx-icon--name-sort rcx-icon rcx-css-4pvxx3"
              >
                
              </i>
            </button>
            <button
              class="rcx-box rcx-box--full rcx-navbar-item rcx-button--small-square rcx-button--square rcx-button--icon rcx-button  rcx-button-group__item"
              title="search"
              type="button"
            >
              <i
                aria-hidden="true"
                class="rcx-box rcx-box--full rcx-icon--name-magnifier rcx-icon rcx-css-4pvxx3"
              >
                
              </i>
            </button>
            <button
              aria-expanded="false"
              aria-haspopup="true"
              class="rcx-box rcx-box--full rcx-button--small-square rcx-button--square rcx-button--icon rcx-button  rcx-button-group__item"
              id="react-aria-3"
              title="create new"
              type="button"
            >
              <i
                aria-hidden="true"
                class="rcx-box rcx-box--full rcx-icon--name-pencil-box rcx-icon rcx-css-4pvxx3"
              >
                ⦑
              </i>
            </button>
          </div>
          <hr
            class="rcx-box rcx-box--full rcx-navbar-divider rcx-divider--vertical rcx-divider"
          />
          <div
            class="rcx-button-group rcx-button-group--align-start rcx-navbar-group"
            role="toolbar"
          >
            <button
              class="rcx-box rcx-box--full rcx-navbar-item rcx-button--small-square rcx-button--square rcx-button--icon rcx-button  rcx-button-group__item"
              title="dialpad"
              type="button"
            >
              <i
                aria-hidden="true"
                class="rcx-box rcx-box--full rcx-icon--name-dialpad rcx-icon rcx-css-4pvxx3"
              >
                
              </i>
            </button>
            <button
              class="rcx-box rcx-box--full rcx-navbar-item rcx-button--small-square rcx-button--square rcx-button--icon rcx-button  rcx-button-group__item"
              title="live"
              type="button"
            >
              <i
                aria-hidden="true"
                class="rcx-box rcx-box--full rcx-icon--name-live rcx-icon rcx-css-4pvxx3"
              >
                
              </i>
            </button>
            <button
              class="rcx-box rcx-box--full rcx-navbar-item rcx-button--small-square rcx-button--square rcx-button--icon rcx-button  rcx-button-group__item"
              title="phone-disabled"
              type="button"
            >
              <i
                aria-hidden="true"
                class="rcx-box rcx-box--full rcx-icon--name-phone-disabled rcx-icon rcx-css-4pvxx3"
              >
                
              </i>
            </button>
            <button
              class="rcx-box rcx-box--full rcx-navbar-item rcx-button--small-square rcx-button--square rcx-button--icon rcx-button  rcx-button-group__item"
              title="message-disabled"
              type="button"
            >
              <i
                aria-hidden="true"
                class="rcx-box rcx-box--full rcx-icon--name-message-disabled rcx-icon rcx-css-4pvxx3"
              >
                
              </i>
            </button>
          </div>
        </span>
        <span
          class="rcx-navbar-section"
        >
          <div
            class="rcx-button-group rcx-button-group--align-start rcx-navbar-group"
            role="group"
          >
            <button
              class="rcx-box rcx-box--full rcx-navbar-item rcx-button--small-square rcx-button--square rcx-button--icon rcx-button  rcx-button-group__item"
              title="preferences"
              type="button"
            >
              <i
                aria-hidden="true"
                class="rcx-box rcx-box--full rcx-icon--name-cog rcx-icon rcx-css-4pvxx3"
              >
                
              </i>
            </button>
            <button
              aria-expanded="false"
              aria-haspopup="true"
              class="rcx-box rcx-box--full rcx-button--small-square rcx-button--square rcx-button--icon rcx-button  rcx-button-group__item"
              id="react-aria-5"
              title="profile"
              type="button"
            >
              <figure
                class="rcx-box rcx-box--full rcx-avatar rcx-avatar--x28"
              >
                <img
                  alt="user avatar"
                  class="rcx-avatar__element rcx-avatar__element--x28"
                  src="data:image/jpeg;base64,/9j/4AAQSkZJRgABAQAAAQABAAD/2wBDAAgGBgcGBQgHBwcJCQgKDBQNDAsLDBkSEw8UHRofHh0aHBwgJC4nICIsIxwcKDcpLDAxNDQ0Hyc5PTgyPC4zNDL/2wBDAQkJCQwLDBgNDRgyIRwhMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjL/wAARCAAoACgDASIAAhEBAxEB/8QAGwAAAgIDAQAAAAAAAAAAAAAAAAcEBgIDBQj/xAAuEAACAQQAAwcEAQUAAAAAAAABAgMABAUREiExBhMUIkFRYQcWcYGhFTJSgpH/xAAYAQADAQEAAAAAAAAAAAAAAAACAwQBAP/EAB4RAAIBBQEBAQAAAAAAAAAAAAABAgMREiExE0HR/9oADAMBAAIRAxEAPwBuXuIkhBuMe5ib/AHQP49q4L3mLitryTLTSpOiHQI5k/HzXa/qbFOEudVTu1dumWvcTaNCZYZ7vU6g6LxqjOU/24dfs1Ouh9FnkMpd3Reeyx83hAxZZEhkdV9/MBrX71WGPvJcqrJBGveKATtuXXqNU0pu02bTHXD/AGvJAluyxxRd6F4x00o+NdKoVrjbzJdvVe1t5cVLc2ck8qjnohgpPtz2v7G6JtPQ2VJwjlcw+37mchpnK6GtIuv5NFWeTsLNPvxWTvpfjvOEfwKKzEVkSct2vscS/BIzSN0YRkeX81UpPqO8masJETu7OOccY4dswYFQeftv096XV5knuJGdm2T1+agvMXj8jEaHX905QihabvcbuS7X566mLWLwSY8PuRnk/u4eZ0deTl71Ef6hY+0yM88TzeNZY4luYwpVYyduOfrvhPTnr0pXSX9y5mCsyJMdyxxvwq599em+taItqCSNc90ChvZRUruUcT0JiO18Elpk7t8v41LWzacxkBSuvjQ/FFJayjDWrCTepAQ2vUH0oo/Jk3ovpwJJeVCP5CN+lFFaaMqy+nAyuChvrTI2kN9JAsi2ZOy4IBHMnkSCP+iqBexSWdxLazoUljJVlPUH2oorkV10pRc7b1zXb/hZOzuJvM86QWEXeELxOzHSIPcmiiiunVlF2RNTpRkrs//Z"
                />
              </figure>
            </button>
          </div>
        </span>
      </nav>
    </div>
  </div>
</body>
`;
